[precompile] Support external data for serialization.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171156
* __->__ #170846

Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path", external_data={...})

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

zhxchen17 committed:
[precompile] Support external data for serialization.

Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path")

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path")

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path")

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path")

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path")

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path")

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste commented on torch/_dynamo/aot_compile.py:
```diff
@@ -60,6 +60,25 @@ def check_compatibility(self) -> None:
 
 class AOTCompilePickler(pickle.Pickler):
+    def __init__(
+        self, external_data: dict[str, Any], *args: Any, **kwargs: Any
```
nit: `object` instead of `Any`

aorenste commented on torch/_dynamo/aot_compile.py:
```diff
@@ -68,26 +87,93 @@ def _() -> Any:
+            """
+            func = obj.__func__
+            method_self = obj.__self__
+            inner_func = getattr(method_self, func.__name__)
```
When is `__name__` a method?

zhxchen17 commented on test/dynamo/test_aot_compile.py:
```diff
@@ -929,6 +937,85 @@ def test_aot_compile_with_redistribute(self):
+        mod.forward = compiled_fn
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Failed to serialize the following objects: \\[SimpleLinearModule",
```
Want to clarify: Is the intention here around giving user the original location of definition (like literally `mod = ...`) in the error message, or giving user a way to know how to find this external data?

Asking because I feel it'd be complicated implement the former as we don't have a good way to backtrack variable definitions from a value in Python. 

I feel in practice "finding the right object to mark as external" shouldn't be a major issue for users, because we only mark torch.nn.Module as external objects, and we can only see at most 1 module object per forward run, so in my mind it's fine to only print the module object (probably can be with an id() as well), and user should be able to find the latest module object used for torch.compile().

I'm also open to other ideas like:
- Return the unhandled objects from save_compiled_function when external_data argument doesn't have it, so that user can add logging for these unhandled objects after `save()` call
- Return something similar to Dynamo source to user like `foo.__closure__[0].cell_contents`, but this doesn't seem a well formed message to users as well.

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path", external_data={...})

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

aorenste reviewed (approved):
I'm satisfied with the responses to my questions - the rest are nits up to the author...

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path", external_data={...})

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path", external_data={...})

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 commented:
@pytorchbot merge

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path", external_data={...})

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

zhxchen17 committed:
Update on "[precompile] Support external data for serialization."


Under some corner cases, users don't directly pass us a free function
or module.forward. Instead, there're multiple levels of wrappers and
captured closures applied to module.forward, so we end up serialzing quite
a few captured data from the compiled function's closure. This is not great
when things like torch.nn.Module end up being captured in function's closure,
and we shouldn't serialize the whole module in this case.

To solve this issue, we specially handle the common cases when the following
objects are being saved:
1. torch.nn.Module
2. Nested functions.

When these data are being serialized, we will mark these as "external data"
since they may contain unserialzable and untrackable states which is
usually easier to be maintained from user side. So the a call to save_compiled_function()
looks like the following:
```
result: AOTCompileSaveResult = compiled_fn.save_compiled_function("/path", external_data={...})

```

On the loading side, user are supposed to provide the external data dict. If any key
is missing, load_compiled_function will throw an error.
```
torch.compiler.load_compiled_function(f, external_data={...})
```

Note that it's not ideal for user to compile a typical function like this, so this
is only meant for power users for whom rewriting model is harder than maintaining
this data map.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
